massa-pos-worker/src/worker.rs

```diff
@@ -61,23 +58,22 @@ impl SelectorThread {
                 initial_rolls,
             };
 
-            this.run(stop_flag)
+            this.run()
         })
     }
 
     /// Check if cycle info changed or new and compute the draws
     /// for future cycle.
     /// # Arguments
     /// * `cycle_info`: a cycle info with roll counts, seed, etc...
-    fn run(mut self, stop_flag: Arc<AtomicBool>) -> PosResult<()> {
+    fn run(mut self) -> PosResult<()> {
         loop {
-            if stop_flag.load(std::sync::atomic::Ordering::Relaxed) {
-                break;
-            }
             let input_data = self.input_data.clone();
             let mut data = input_data.1.lock();
-            if let Some(cycle_info) = data.pop_front() {
-                self.draws(cycle_info)?
+            match data.pop_front() {
+                Some(Command::CycleInfo(cycle_info)) => self.draws(cycle_info)?,
+                Some(Command::Stop) => break,
+                None => {}
             }
```

Comment:
You could do something like:
```rust
let cycle_info = {
    // do stuff with the lock and condvar
};
self.draws(cycle_info);
```
so as not to hold the lock while performing the draw(unless the draw should be part of the criticial section)

Reply:
done